Fix 2601

kinnison opened the pull request:
Previously we attempted to resolve a situation where a toolchain which had a wildcarded component installed as target-specific due to profiles would break things.  Sadly it seems we missed one other mechanism by which such a component installation might occur - namely if someone requests a wildcarded component as an *additional* component during toolchain installation.

This PR resolves this by detecting such components during the building of the dist changeset.  As such, it fixes #2601 

@eddyb When this has built, assuming a green CI, if you're logged into Github then you should be able to download an artifact from the relevant CI run and try it out.  Would you mind checking if this fixes your problem?

kinnison committed:
tests: Add test exposing 2601

Signed-off-by: Daniel Silverstone <dsilvers@digital-scurf.org>

kinnison committed:
dist: Identify requested additional component wildcards

Sometimes someone might choose an additional component to add
during an install/update which is wildcarded.  We need to ensure
that we detect this and request the install with the wildcard
otherwise we can hit situations such as that identified in #2601

Signed-off-by: Daniel Silverstone <dsilvers@digital-scurf.org>

eddyb commented:
I can't use the artifact myself directly, but I can do what I did in https://github.com/NixOS/nixpkgs/pull/106443, only just locally, and pointing to your branch, then I get a "distro `rustup`" that I can do all my testing with.

I'll have to do that tomorrow though, but thanks for the fix!

eddyb commented:
I can confirm the sequence from #2601 behaves correctly, *but* that's because it starts by removing `nightly`.

But if I have `rust-src-x86_64-unknown-linux-gnu` in `lib/rustlib/components` from an older `rustup`, this PR makes *both* methods to install a component fail in the manner described in https://github.com/rust-lang/rustup/issues/2601#issuecomment-741592459 (to reproduce, run the first two commands in that comment with an older `rustup`).

So as I first suspected, this PR makes `rustup toolchain install ... --component rust-src` behave like `rustup component add rust-src`, which is an improvement, *but* `rustup component add rust-src` failed when `rust-src-x86_64-unknown-linux-gnu` was installed, and probably requires work elsewhere to trigger the self-repair.

Interestingly enough, installing *any other* component triggers the self-repair, but for `rust-src` I guess it triggers too late?

kinnison commented:
Oh crap yes, if you're already broken then it won't trigger the repair. Hmm, I'll need to decide how to handle this more cleanly.  This was mostly to not break in the first place I hope.

kinnison committed:
Update::build_final_component_list: Fix self-healing

When self-healing from a bad rust-src install as a side effect
of someone running `rustup component add rust-src` we would accidentally
double-add the rust-src component resulting in a conflict.

This fixes that.

Signed-off-by: Daniel Silverstone <dsilvers@digital-scurf.org>

kinnison commented:
I think this was being over-eager when self-healing rust-src misinstallation.  We ended up double-installing rust-src which is what caused the actual conflict listed.  Assuming CI goes green, would you give this branch a test for you.  I think it's good now.

eddyb commented:
My bad, I haven't had a chance to look at this yet, it slipped through the cracks.

kinnison commented:
@eddyb That's quite alright, everyone's busy :D  Just let me know when you can give it a go please?

eddyb commented:
The issue I described in https://github.com/rust-lang/rustup/pull/2602#issuecomment-742738157 (with a a toolchain from an older `rustup`) appears to be fixed, and the same-`rustup` behavior is still fixed.

Again apologies for the (extra one week) delay, I've been having a hard time doing software things lately.

kinnison commented:
No need to apologise, this time of year is tough for most people to get time, let alone to get time to deal with small side issues.

Thank you for checking it out, I'm going to merge this then.
